[게시판 회원 기능] 쿠기 미션 제출합니다.

men7627 opened the pull request:
안녕하세요 포비 ! 
이번 미션은 우테코에 참여하면서 가장 힘들기도 하면서도, 가장 열심히 했고, 많은 걸 배웠던 미션이었던 것 같습니다.

미션 구현이 더디어도 최대한 많은 것들을 공부하고 적용하려고 노력했습니다.
- lombok
- 중첩 DTO 클래스
- Valid annotation
- ControllerAdvice, ExceptionHandler
- Controller, Service 분리
... 
등 새로운 것들을 많이 적용해보는 경험이 있었습니다.

저의 코드 중 마음에 들지 않는 부부은 각 요청에 대한 메서드에서 session이 있는지 검사를 하는 코드가 중복이 많은데 이를 없애는 방벙을 찾지 못하여서 아쉬웠습니다.

어려웠던 점은 Test Code 였습니다.
첫번째는 Session에 대한 테스트 코드 였는데 , Session이 필요한 경우, 필요하지 않는 경우를 분리해서 테스트 하려고 하는데 방법을 잘 모르다보니 힘들 었습니다.
결국에는 Cookie를 직접 만들어서 주입하는 방식으로 구분하여 테스트하였습니다.

그리고, 최종 미션 구현 및 테스트 완료 후 페어의 노트북에서 작업한 것을 머지하여 저의 노트북에서 전체 테스트를 돌려보았는데, 분명 페어 노트북에서는 다 성공하던 것이 저의 노트북에서는 일부 실패하는 것을 발견하였습니다.
자세히 살펴보니, 각 테스트 클래스에 대한 실행 순서가 달랐습니다. 그래서 각 테스트 클래스는 다른 테스트 클래스에 영향을 줄 수 있다는 것을 알았습니다. 어떤 문제인지 디버그를 통해서 살펴보았습니다.
발견한 원인은 ServiceTest에서 DB에 접근하여 저장을 하고 테스트를 수행하고 난 후, Controller에서 테스트를 할 때 증가 된 Auto Increment 값으로 테스트를 하게 된다는 것을 알게 되었습니다.
(기존에는 다른 테스트 클래스가 영향을 주는지 몰라서 Controller에서 Id값을 전역으로 1로 초기화하고 수행하였습니다.)
결국 Test코드를 전체 패키지로 돌렸을 때 Spring Boot를 한번 실행해서 다 돌리게 되고 그에 따라서 H2DB도 한번 실행해서 전체를 돌린다는 사실을 깨닫게 되었습니다.

이를 해결하기 위해서 각 테스트 클래스에 DirtiesContext annotation을 작성하여, 각 테스트 클래스를 실행할 때 강제로 서버를 재실행하게 하였습니다.
처음에는 전역 id값을 없애는 방향으로 수정하려 했으나 쉽지 않아서 서버를 재실행하는 방법을 택하게 되었습니다.

잘 부탁드립니다 !

seonghun127 committed:
setting: mission complete

seonghun127 committed:
docs: README.md

1. todo 목록 정리

seonghun127 committed:
feat: build.gradle

1. devtools 의존성 추가

seonghun127 committed:
feat: ArticleController Class

1. Article 도메인 객체 구현
2. Test 추가
3. ArticleRepository 저장기능 구현

seonghun127 committed:
feat: layout

1. build.gradle 의존성 추가
2. 공통 레이아웃 추출 및 적용

seonghun127 committed:
feat: ArticleDto Class

1. Create 용도의 ArticleDto 구현
2. ArticleDto를 사용하도록 Controller와 Repository 변경

seonghun127 committed:
refactor: Article class

1. ArticleDto를 사용하기 위한 생성자 추가
2. id 자동 증가를 위한 static field 추가

seonghun127 committed:
feat: ArticleController class

1. 조회 기능 구현

seonghun127 committed:
style: css 및 templates 파일 코드 포메팅

seonghun127 committed:
fix: app.css

1. 파일 오류 수정

seonghun127 committed:
test: ArticleRepository class

1. 조회 및 전체조회 테스트

seonghun127 committed:
fix: layout

1. thymeleaf-layout 버전에 따른 오류 수정

seonghun127 committed:
fix: ArticleController class

1. createArticle 리다이렉팅 주소 수정

seonghun127 committed:
feat: ArticleController class

1. 게시물 수정 기능 구현
2. ArticleDto.Update 생성
3. ArticleRepository에 update 기능 추가
4. update를 위해 Article 도메인 객체에 생성자 추가

codeanddonuts committed:
init

codeanddonuts committed:
feat: Article POJO

codeanddonuts committed:
feat: Article repository

codeanddonuts committed:
feat: router

seonghun127 committed:
feat: article-edit.html

1. 생성 화면과 수정 화면에 따른 form 경로 및 메서드 조정
2. article.html에 하드코딩된 url 변경

seonghun127 committed:
feat: ArticleController class

1. 삭제 기능 구현
2. ArticleRepository delete 구현

seonghun127 committed:
feat: article.html

1. delete 버튼 구현

seonghun127 committed:
refactor: layout

1. header 공통부분 추출

seonghun127 committed:
fix: ArticleRepository class

1. 수정시 새 게시물이 추가되는 오류 수정

seonghun127 committed:
style: 전체 코드 포매팅

codeanddonuts committed:
fix: article delete

codeanddonuts committed:
fix: put method

codeanddonuts committed:
test: controller tests

seonghun127 committed:
fix: article-edit.html

1. 수정 페이지에서 글 내용이 나오지 않는 오류 수정

seonghun127 committed:
fix: article.html

1. 예시로 설정된 게시글 내용 삭제
2. 작성된 게시글 내용이 나오게끔 수정

codeanddonuts committed:
test: repo tests

seonghun127 committed:
refactor: custom head 적용 방법 변경

seonghun127 committed:
refactor: fragment 폴더 생성

codeanddonuts committed:
feat: live reloading and auto refresh support

seonghun127 committed:
refactor: layout.html

1. main-head를 담는 th:replace 태그 추가
2. main-body를 담는 th:replace 태그 추가

codeanddonuts committed:
fix : layout

codeanddonuts committed:
doc: README.md

seonghun127 committed:
refactor: ArticleController class

1. 필드 의존 주입에서 생성자 의존 주입으로 변경

seonghun127 committed:
refactor: NotFoundArticleException class 추가

seonghun127 committed:
refactor: ArticleRepository class

1. find() 메서드의 이름 수정 : findById()

seonghun127 committed:
style: 코드 포매팅

seonghun127 committed:
refactor: Article class

1. set 메서드 제거

seonghun127 committed:
fix: ArticleRepositoryTest 클래스에서 깨지는 테스트 코드 수정

seonghun127 committed:
refactor: ArticleController class

1. @AllArgsContructor 를 @RequiredArgsContructor 어노테이션으로 변경
2. articleRepository 필드를 final 변경
- 불변셩을 유지하기 위해

men7627 committed:
fix : ArticleControllerTest

seonghun127 committed:
feat: Article class

1. 커스터마이즈한 Builder 패턴 추가

seonghun127 committed:
refactor: Article class에 lombok @Builder 적용

seonghun127 committed:
feat: ArticleController class

1. 게시물 추가 테스트 코드 구현

seonghun127 committed:
style: ArticleRepositoryTests class 이름 변경

seonghun127 committed:
fix: Article class

1. updateArticle() 메서드 추가

seonghun127 committed:
refactor: 중복 head 부분 제거

seonghun127 committed:
feat: ArticleController class

1. modelMapper 의존성 주입
2. modelMapper 클래스를 빈으로 등록
3. ArticleDto 클래스에 Response 중첩클래스 추가
4. ArticleController에서 article(s) 를 반환하는 모든 API의 반환 형태를 ArticleDto.Response 타입으로 변경

seonghun127 committed:
style: 코드 포맷팅

seonghun127 committed:
refactor: ArticleController class

1. updateArticle() 메서드 리턴 방식 변경
- 리다이렉트에서 포워딩 방식으로 변경한다
2. 그에 따른 인수 테스트 방식 수정
- 수정된 값을 확인하는 테스트로 수정한다

seonghun127 committed:
style: 주석 제거

seonghun127 committed:
refactor: ArticleRespository class

1. save() 메서드 리턴 타입 void로 변경
2. update() 메서드 리턴 타입 void로 변경

seonghun127 committed:
refactor: ArticleController class

1. createArticle() 메서드 리턴 타입 변경
- 리다이렉트 방식에서 포워딩 방식으로 변경한다

seonghun127 committed:
refactor: ArticleControllerTests class

1. 지속 테스트 가능한 코드로 리펙토링

seonghun127 committed:
style: 코드 포맷팅

seonghun127 committed:
fix: ArticleController class

1. createArticle(), updateArticle() api 리턴 타입 변경
- 포워드 방식에서 리다이렉트 방식으로 변경한다

2. ArticleRepository 클래스에서 save(), update() 메서드의 리턴 타입 변경
- void에서 long으로 변경한다

3. 테스트 코드 수정

seonghun127 committed:
style: 코드 포맷팅

seonghun127 committed:
[게시글 생성/조회/수정/삭제] 이바 미션 제출합니다. (#19)

* setting: mission complete

* docs: README.md

1. todo 목록 정리

* feat: build.gradle

1. devtools 의존성 추가

* feat: ArticleController Class

1. Article 도메인 객체 구현
2. Test 추가
3. ArticleRepository 저장기능 구현

* feat: layout

1. build.gradle 의존성 추가
2. 공통 레이아웃 추출 및 적용

* feat: ArticleDto Class

1. Create 용도의 ArticleDto 구현
2. ArticleDto를 사용하도록 Controller와 Repository 변경

* refactor: Article class

1. ArticleDto를 사용하기 위한 생성자 추가
2. id 자동 증가를 위한 static field 추가

* feat: ArticleController class

1. 조회 기능 구현

* style: css 및 templates 파일 코드 포메팅

* fix: app.css

1. 파일 오류 수정

* test: ArticleRepository class

1. 조회 및 전체조회 테스트

* fix: layout

1. thymeleaf-layout 버전에 따른 오류 수정

* fix: ArticleController class

1. createArticle 리다이렉팅 주소 수정

* feat: ArticleController class

1. 게시물 수정 기능 구현
2. ArticleDto.Update 생성
3. ArticleRepository에 update 기능 추가
4. update를 위해 Article 도메인 객체에 생성자 추가

* feat: article-edit.html

1. 생성 화면과 수정 화면에 따른 form 경로 및 메서드 조정
2. article.html에 하드코딩된 url 변경

* feat: ArticleController class

1. 삭제 기능 구현
2. ArticleRepository delete 구현

* feat: article.html

1. delete 버튼 구현

* refactor: layout

1. header 공통부분 추출

* fix: ArticleRepository class

1. 수정시 새 게시물이 추가되는 오류 수정

* style: 전체 코드 포매팅

* fix: article-edit.html

1. 수정 페이지에서 글 내용이 나오지 않는 오류 수정

* fix: article.html

1. 예시로 설정된 게시글 내용 삭제
2. 작성된 게시글 내용이 나오게끔 수정

* refactor: custom head 적용 방법 변경

* refactor: fragment 폴더 생성

* refactor: layout.html

1. main-head를 담는 th:replace 태그 추가
2. main-body를 담는 th:replace 태그 추가

* refactor: ArticleController class

1. 필드 의존 주입에서 생성자 의존 주입으로 변경

* refactor: NotFoundArticleException class 추가

* refactor: ArticleRepository class

1. find() 메서드의 이름 수정 : findById()

* style: 코드 포매팅

* refactor: Article class

1. set 메서드 제거

* fix: ArticleRepositoryTest 클래스에서 깨지는 테스트 코드 수정

* refactor: ArticleController class

1. @AllArgsContructor 를 @RequiredArgsContructor 어노테이션으로 변경
2. articleRepository 필드를 final 변경
- 불변셩을 유지하기 위해

* feat: Article class

1. 커스터마이즈한 Builder 패턴 추가

* refactor: Article class에 lombok @Builder 적용

* feat: ArticleController class

1. 게시물 추가 테스트 코드 구현

* style: ArticleRepositoryTests class 이름 변경

* fix: Article class

1. updateArticle() 메서드 추가

* refactor: 중복 head 부분 제거

* feat: ArticleController class

1. modelMapper 의존성 주입
2. modelMapper 클래스를 빈으로 등록
3. ArticleDto 클래스에 Response 중첩클래스 추가
4. ArticleController에서 article(s) 를 반환하는 모든 API의 반환 형태를 ArticleDto.Response 타입으로 변경

* style: 코드 포맷팅

* refactor: ArticleController class

1. updateArticle() 메서드 리턴 방식 변경
- 리다이렉트에서 포워딩 방식으로 변경한다
2. 그에 따른 인수 테스트 방식 수정
- 수정된 값을 확인하는 테스트로 수정한다

* style: 주석 제거

* refactor: ArticleRespository class

1. save() 메서드 리턴 타입 void로 변경
2. update() 메서드 리턴 타입 void로 변경

* refactor: ArticleController class

1. createArticle() 메서드 리턴 타입 변경
- 리다이렉트 방식에서 포워딩 방식으로 변경한다

* refactor: ArticleControllerTests class

1. 지속 테스트 가능한 코드로 리펙토링

* style: 코드 포맷팅

* fix: ArticleController class

1. createArticle(), updateArticle() api 리턴 타입 변경
- 포워드 방식에서 리다이렉트 방식으로 변경한다

2. ArticleRepository 클래스에서 save(), update() 메서드의 리턴 타입 변경
- void에서 long으로 변경한다

3. 테스트 코드 수정

* style: 코드 포맷팅

men7627 committed:
fix : ArticleRepositoryTest

men7627 committed:
fix : ArticleControllerTest

men7627 committed:
fix : ArticleControllerTest

men7627 committed:
fix : way of put method using

men7627 committed:
fix : modify url

seonghun127 committed:
fix: Spring Data Jpa 사용

seonghun127 committed:
docs: 기능 구현 목록 정리

seonghun127 committed:
docs: h2 기본 설정

seonghun127 committed:
docs: 기능 구현 목록 정리

seonghun127 committed:
feat: 회원가입 요청 페이지 이동 및 저장

1. 회원 가입 페이지 이동 api(renderSignUpPage())를 구현한다
2. 회원 정보를 저장하는 api(createUser())를 구현한다

seonghun127 committed:
merge local branch into fork repository branch

seonghun127 committed:
feat: 회원등록 유효성 검사

seonghun127 committed:
fix: 회원가입 api response 방식 변경

1. createUser() api의 리턴 방식을 리다이렉트로 변경한다
- 회원가입이 성공하면 "/login"으로 리다이렉트 된다
- 회원가입이 실패하면 "/signup"으로 리다이렉트 된다

seonghun127 committed:
feat: 회원 조회 기능 구현

1. 회원을 조회하는 api(readUsers())를 구현한다
2. service 계층에서 dto와 domain 객체의 mapping 처리한다

seonghun127 committed:
fix: 페이지 링크 수정

seonghun127 committed:
feat: 서버에 예외처리 추가

1. email, password, name에 형식 조건을 추가한다
2. ControllerAdvice로 exception handler를 관리한다

seonghun127 committed:
feat: UserControllerTest class

seonghun127 committed:
docs: 기능 구현 목록 정리

seonghun127 committed:
refactor: 서버에서 예외처리 리팩토링

1. 커스텀 예외를 추가한다
- 중복, 형식(이메일, 이름, 비밀번호)
2. ControllerAdvice와 ExceptionHandler를 적용한다
3. Response Status를 설정한다

seonghun127 committed:
feat: 로그인, 로그아웃 기능 구현

1. Login 전용 DTO를 추가한다 (UserDTO.Login)
2. 로그인 성공 시 Session을 생성한다
3. 로그인 유무에 따라 페이지를 분기 이동한다

seonghun127 committed:
fix: 페이지 요청 방식 수정

1. get 요청 시 화면, url 일관성 주기 위해 redirect방식으로 변경한다

seonghun127 committed:
docs: 기능 구현 목록 정리

seonghun127 committed:
feat: MyPage 이동 기능 구현

1. 마이페이지로 이동하는 api(renderMYPage())를 구현한다
2. 로그인 유무에 따라 이동 페이지를 결정한다

seonghun127 committed:
fix: 로그인이 안되있을 시 로그아웃 기능 수정

1. session 정보가 존재할때만 로그아웃을 수행한다

seonghun127 committed:
docs: 기능 목룩 구현 정리

seonghun127 committed:
feat: 회원 정보 수정 기능 추가

1. 회원 정보 수정 api(updateUser())를 추가한다
2. My Page로 이동하는 api(renderMypage())를 수정한다
- session 정보를 확인하는 로직을 삭제한다
3. Edit My Page로 이동하는 api(renderEditMypage())를 추가한다
- renderMypage()에서 처리하던 session 관리 로직을 추가한다
4. UserDto.Reponse 클래스에 id 필드를 추가한다
5. Service 계층에 update() 메서드를 추가한다

seonghun127 committed:
fix: 회원가입, 회원수정 시 예외 처리 코드 분류

1. InvalidSignUpFormException을 구현한다
2. InvalidEditFormException을 구현한다

seonghun127 committed:
feat: 회원 탈퇴 기능 구현

1. 회원 정보를 삭제하는 api(deleteUser()) 추가한다
2. 테스트 코드 @AfterEach에 삭제 api를 호출하여 전체 테스트가 가능하게 한다

seonghun127 committed:
feat: UserService 테스트 코드 추가

seonghun127 committed:
fix: UserService 테스트 코드 수정

seonghun127 committed:
refactor: ArticleService 계층 분리

seonghun127 committed:
style: 코드 포맷팅

seonghun127 committed:
docs: data.sql 추가

men7627 committed:
Revert "[게시글 생성/조회/수정/삭제] 쿠기 미션 제출합니다. (#12)"

This reverts commit 20ede201c6d9eddc2998c8ab75cba3a3190fbe03.

men7627 committed:
Merge remote-tracking branch 'eva/seonghun127' into men7627

men7627 committed:
merge

men7627 committed:
merge

seonghun127 committed:
fix: 테스트 코드 오류 수정

1. data.sql 파일 삭제

men7627 committed:
data.sql 삭제

men7627 committed:
..

men7627 committed:
fix: 각 테스트 클래스에 DirtiesContext추가

pobiconan reviewed (changes requested):
정말 다양한 기술을 적용하고 시도했네요. 💯 
일주일 사이에 이 모든 것을 적용했다는 것이 믿기지 않을 정도로..
단, 세션 체크에 대한 중복도 많고, 잘못 처리하는 부분이 보이네요. 
세션에 추가한 "user" 데이터를 기준으로 존재 여부를 체크하고, 삭제하는 방식으로 구현해 보세요.

men7627 committed:
fix : compile을 implementation으로 수정

men7627 committed:
fix : HttpServletRequest를 HttpSession으로 변경

men7627 committed:
fix : login, signup 페이지 메인로고 url 수정

men7627 committed:
fix : 생성자를 annotation으로 대체

men7627 committed:
fix : ArticleInterceptor 구현

1. ArticleController 로그인 여부 체크 중복을 제거한다

men7627 committed:
fix : handelInvalidEditFormException 오류 수정

1. optional 값을 get()을 통해 가져온다

men7627 committed:
fix : 테스트 body 입력 값 수정

men7627 committed:
fix : Interceptor 구현

1. ArticleInterceptor, UserInterceptor를 구현한다
2. 각 Controller에 있는 session의 user 존재 여부를 검사하는 중복을 제거한다
3. 각 Interceptor에 대한 테스트를 각 ControllerTest에 추가한다

men7627 committed:
fix : DTO에 유효성 annotation 추가

men7627 commented:
피드백을 반영하면서 Interceptor에 많은 시간을 썼습니다.
Interceptor로 로그인에 대한 처리를 최대한 모든 경우를 다 처리할려다 보니 어려움을 느꼈습니다.
각 경우를 나누는 것도 생각할 것이 많아서 쉽지 않았고, 모든 경우를 코드로 구현하려다 보니 if문이 많이 생기고 복잡해졌습니다.
원하는 방식으로 구현에는 성공하여 각 Controller에 Interceptor에 대한 테스트도 추가했습니다.
구현은 하였으나 수정해야 할 부분이 많은 것 같습니다.

men7627 committed:
fix : AbstractTest 구현

pobiconan reviewed (approved):
피드백 반영하느라 수고했네요. 💯 
그런데 Interceptor로 처리하는 경우 기능이 추가될 수록 변경사항이 계속해서 발생하는 단점이 있는 것 같네요.
다음 단계 진행하면서 다른 방법이 있는지 찾아보면 좋을 것 같아요.
